### ISSUE TITLE
dart2js: diet parse hides common syntax errors

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#24232]
``` dart
const a = 1
const b = 2;

main() => b;
```

Note the missing semicolon after `1`. Today dart2js compiles this and produces this output:

```
> dart2js syn_err.dart
syn_err.dart:4:11:
Warning: Cannot resolve 'b'.
main() => b;
          ^
Dart file (syn_err.dart) compiled to JavaScript: out.js
```

This should be a compile-time error.

**Comment by sigmundch** [comment]
FWIW, the issue here relates to diet parsing - the code is treated like:

```dart
const a = #; // when needed, parse the rhs from "1 const b = 2".
main() => b;
```

If you for example change the code and use `a` instead of `b` in the main function:

```dart
const a = 1 b = 2;
main() => a;
```

, you'll get the expected error:
```
debug.dart:1:9:
Error: Expected ';' after this.
var a = 1 b = 2;
        ^
```

@johnniwinther @peter-ahe-google - you are probably more familiar with this, so I got a few questions for you:
* Is there something that can be done for this, or is it a limitation of diet parsing we need to live with?

* The warning is a good hint, but the context where I discovered this was in a large app were we were omitting warnings from dependent packages, so the hint `cannot resolve 'b'` was not visible by default. Maybe we need some levels of warnings and show some important warnings regardless?

* Is diet-parsing of initializers worth it? I'm guessing we gain a lot from skipping classes and top-level functions, but how much are we gaining from skipping top-levels initializers like the one above?
